### PR TITLE
Dashboard

### PR DESCRIPTION
**Pull request by Miclin1024**
Done with the staff panel as well as the dashboard summary card

### CODE REVIEW

File: accounts/views.py [hunk 1]
```diff
@@ -30,7 +30,16 @@ def manage_library(request):
 
 @login_required
 def dashboard(request):
+    if request.GET.get('endsession', False):
```
Comment (chenyx512) [inline accounts/views.py:33]: why is this here but not in learning_process

File: templates/accounts/dashboard.html [hunk 2]
```diff
@@ -20,6 +20,16 @@ <h5 class="col-12 col-md-8 align-self-center text-center" id="sidebar-info" >{{
                     <small>{{ user.email }}</small>
                 </h5>
             </div>
+            {% if user.is_superuser == 1 %}
```
Comment (chenyx512) [inline templates/accounts/dashboard.html]: '== 1' is unnecessary

File: learning/views.py [hunk 3]
```diff
@@ -23,6 +23,15 @@ def display_character(request, character_pk):
     except ObjectDoesNotExist:
         character = Character.objects.filter(pk__gt=character_pk).first()
         return redirect('display_character', character_pk=character.pk)
+    try:
+        status = request.session['is_learning']
+    except KeyError:
+        status = False
```
Comment (chenyx512) [inline learning/views.py]: simplify this with `request.session.get('is_learning', False)`

File: learning/views.py [hunk 4]
```diff
@@ -23,6 +23,15 @@ def display_character(request, character_pk):
...
+    if status == True:
+        learning_process(request)
+    else:
+        request.user.last_study_vocab_count = 0
```
Comment (chenyx512) [inline learning/views.py:32]: After changing a model you would have to call its save methods to write it into DB,
here it would be `request.user.save()`

File: templates/accounts/dashboard.html [hunk 5]
```diff
@@ -31,6 +41,32 @@ <h5 class="col-12 col-md-8 align-self-center text-center" id="sidebar-info" >{{
             </a>
         </div>
         <div class="offset-1 offset-sm-2 col-11 col-sm-10">
+            {% if active == "Admin"%}
```
Comment (chenyx512) [inline templates/accounts/dashboard.html]: here don't use the active thing, instead, say '{% if user.is_staff %}'. Also you don't need to have a sperate thing for superusers, since all superusers know the urls to those links... 
So this staff panels need only be implemented in frontend, please delete the things in views...

File: accounts/views.py [hunk 6]
```diff
@@ -30,7 +30,16 @@ def manage_library(request):
...
+        request.session['is_learning'] = False
     active = request.GET.get('active', 'Dashboard')
+
+    if active == 'Admin' and request.user.is_superuser != True: 
+        HttpResponseRedirect(reverse('dashboard')+"?active=Dashboard")
+
+    if active == 'Staff' and request.user.is_staff != True:
+        HttpResponseRedirect(reverse('dashboard')+"?active=Dashboard")
+
```
Comment (chenyx512) [inline accounts/views.py]: above is unnecessary, those should be implemented in frontend

File: .gitignore [hunk 7]
```diff
@@ -7,6 +7,7 @@ __pycache__
 .idea
 *.log
 *.pot
+.fuse_hidden*
```
Comment (chenyx512) [inline .gitignore]: I added this line to ignore temporary files in linux and mac

File: learning/views.py (hunk 4 above)
Comment (chenyx512) [reply learning/views.py:32]: also why do we need to reset `last_study_vocab_count` if user is not learning? User may be just viewing the character. The character display view/template are also used later for dictionary function.

File: learning/views.py [hunk 8]
```diff
@@ -23,6 +23,15 @@ def display_character(request, character_pk):
...
+        start_learning(request, 10)
```
Comment (chenyx512) [inline learning/views.py:33]: here `start_learning()` should be called in the `urls.py` if user clicks start learning in index page. The reason the start learning button directs the user here is because the learning logic isn't finished...

File: learning/views.py [hunk 9]
```diff
@@ -53,9 +66,17 @@ def learning_process(request):
     delta_time = timezone.now() - request.session['last_record_time']
     request.session['last_record_time'] = timezone.now()
     request.user.last_study_duration += delta_time
+    request.user.last_study_vocab_count += 1
```
Comment (chenyx512) [inline learning/views.py:69]: this view is the general view for learning, which means it will also handle review and relearn. Therefore `last_study_vocab_count` shouldn't be updated here...

**Review by chenyx512 — CHANGES_REQUESTED** [review]

File: templates/accounts/dashboard.html (hunk 5 above)
Comment (Miclin1024) [reply templates/accounts/dashboard.html]: The **active attribute** indicates which of the 4 tabs is activated and therefore shown in the dashboard. The idea is to allow the staff to have access to the normal user dashboard interface, for debug/testing or whatever reasons, while providing a staff tab to give them quick access to the URLs.

Comment (chenyx512) [reply templates/accounts/dashboard.html]: ah here I understand, you are correct

File: templates/accounts/dashboard.html [hunk 10]
```diff
@@ -168,6 +204,35 @@ <h1>
...
+    } else if (strHour - {{ user.last_study_time.hour}} > 0) {
+        document.getElementById("last-study-date").innerHTML = strHour - {{ user.last_study_time.hour }}
+        document.getElementById("last-study-date-unit").innerHTML = "Hours ago"
+    } else if (strMin - {{ user.last_study_time.minute}} > 0) {
+        document.getElementById("last-study-date").innerHTML = strHour - {{ user.last_study_time.minute }}
+        document.getElementById("last-study-date-unit").innerHTML = "Minutes ago"
+    } else {
+        document.getElementById("last-study-date").innerHTML = ""
+        document.getElementById("last-study-date-unit").innerHTML = "Less than a Minute ago"
+    }
```
Comment (chenyx512) [inline templates/accounts/dashboard.html]: for js, remember to put semicolons...
